Add application relation data and event handling

dshcherb opened the pull request:
Add application relation data and event handling

dshcherb commented:
Modified the tests introduced with this PR to use the fake_script approach.

niemeyer commented on op/model.py:
```diff
@@ -186,26 +206,34 @@ def __getitem__(self, key):
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
```
There's actually a way around that. We can cache the time we did the call, and then prevent calling it again within that timeframe. This is safe as long as the lease time doesn't change, and as long as we check the time _before_ we do the call (so the window is guaranteed correct by the ordering).

niemeyer replied:
Ah, and ideally the clock used would be monotonic. How is Python doing in that regard nowadays?

johnsca replied:
[`time.monotonic()`](https://docs.python.org/3/library/time.html#time.monotonic) is in the stdlib (since 3.3, with improvements in 3.5), so it seems like it should be reliable.

stub42 replied:
Before finalizing this, it is worth finding out what happens if leader-set was successfully called before losing leadership. Do the leader-set changes silently roll back? Or does the hook fail? I suspect the hook fails if leadership is lost mid-hook (or it should). Also need to find out if leader-set renews the lease, or if the charm has no way of knowing how old the lease is.

(I don't think any charms worry about it and will just fail, since it involves a netsplit happening when running a hook that takes >30s to run)

dshcherb replied:
For reference, here is the code path from the is-leader hook to to the `ClaimLeader` call in the leadership tracker.

https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/jujuc/is-leader.go#L47
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L55-L68 (`ensureLeader` is called here).
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L103-L113
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/leadership/tracker.go#L109-L112

Based on what we discussed so far, my understanding is that we will not use `leader-set` and instead use `relation-set --app` making the framework depend on Juju 2.7 as a base.

> if leader-set renews the lease

`leader-set` also makes sure that the current unit is a leader by calling `ensureLeader`.

https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L70-L76
The test for leader-set covers the failure scenario so an error is returned when 

Which is also covered by this test scenario:
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader_test.go#L200-L215

> Do the leader-set changes silently roll back?

If `leader-set` returns successfully, it means that all changes were successfully committed to the controller db during its execution and the changes will not be rolled back, i.e. `leader-set` changes get committed immediately but `leader-get` is cached at the unit agent side.

I brought this up as a discussion point during our first sprint and it was decided that we will move to peer application relation data instead to have identical commit semantics for all data bags.

dshcherb replied:
@niemeyer @johnsca @jameinel 

Added another commit to this PR to cover temporary caching of is-leader return values as suggested.

dshcherb replied:
Moved this to a separate PR:

https://github.com/canonical/operator/pull/59

dshcherb commented on op/main.py:
```diff
@@ -116,10 +116,19 @@ def _get_event_args(charm, bound_event):
+        else:
+            remote_app = None
+
+        if remote_app:
```
I think we have to use 2.7 as a base for the framework because of the features such as application relation data, podspec v2, actions/functions for k8s charms.

As far as I understand from the discussions we had up until this point, we will not support `leader-{get,set}` and instead use peer app relation data.

Hence I assume unconditionally here that `JUJU_REMOTE_APP` is available.

Let's have a discussion about this during our call.

dshcherb replied:
> Do we actually have a reason today to limit ourselves to juju 2.7?
> The things mentioned above don't seem relevant for this

Leader data is a major point. Are you suggesting we have one public implementation exposed to the framework users that uses `leader-{get, set}` before 2.7 and peer application relation data afterwards? Besides having to implement things in 2 different ways we would add an upgrade step as well.

dshcherb replied:
Per the discussion during our call:

* I will update the PR to obtain a remote application name from a remote unit name to support earlier versions;
* We will not implement support for leader-{get, set} until somebody asks for it for Juju < 2.7;
   * If we implement it, it will be a separate API and **not** a shim that uses leader-{get, set} before 2.7 and peer app relation data after 2.7 with an upgrade path (in other words, we will not do anything behind a user's back).

niemeyer commented on op/model.py:
```diff
@@ -287,9 +314,9 @@ def relation_list(self, relation_id):
                 raise
 
-    def relation_get(self, relation_id, member_name):
+    def relation_get(self, relation_id, member_name, app):
```
Is that an app or a boolean?  The PR has code that hints it both ways.

niemeyer commented on op/model.py:
```diff
@@ -312,4 +339,13 @@ def config_get(self):
+        The value is cached for the duration of a lease which is 30s in Juju.
+        """
+        if time.monotonic() - self._leader_check_time > self.LEASE_REFRESH_PERIOD or self._is_leader is None:
+            # Record the time before executing is-leader to account for the execution delay and so make the next check to happen earlier.
```
```
# Current time MUST be saved before running is-leader to ensure the cache
# is only used inside the window that is-leader itself asserts.
```

dshcherb replied:
Will address in https://github.com/canonical/operator/pull/59

niemeyer commented on op/model.py:
```diff
@@ -312,4 +339,13 @@ def config_get(self):
+            self._leader_check_time = time.monotonic()
+            self._is_leader = self._run('is-leader')
+
+        return self._is_leader
```
This change must be on its own PR as it's unrelated to the implementation of application relation data and it's a relevant delta in behavior.  This is way too large by itself already.

dshcherb replied:
#59

niemeyer commented on test/test_main.py:
```diff
@@ -43,6 +50,15 @@ class SymlinkTargetError(Exception):
+    def __init__(self, event_type, event_name, relation_id=None, remote_app=None, remote_unit=None):
+        self.event_type = event_type
+        self.event_name = event_name
+        self.relation_id = str(relation_id)
```
Doesn't seem appropriate to stringify it here. It's assuming a specific use case which should be handled when it becomes relevant.

dshcherb replied:
I added it here to cover the following error:
```
            env.update({
                'JUJU_RELATION': rel_name,
                'JUJU_RELATION_ID': event_spec.relation_id,
            })
```
```
Traceback (most recent call last):
  File "/home/dima/src/canonical/juju/operator/test/test_main.py", line 220, in test_multiple_events_handled
    state = self._simulate_event(event_spec, charm_config)
  File "/home/dima/src/canonical/juju/operator/test/test_main.py", line 151, in _simulate_event
    subprocess.check_call([sys.executable, event_hook], env=env)
  File "/usr/lib/python3.7/subprocess.py", line 358, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/usr/lib/python3.7/subprocess.py", line 339, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1462, in _execute_child
    env_list.append(k + b'=' + os.fsencode(v))
  File "/usr/lib/python3.7/os.py", line 810, in fsencode
    filename = fspath(filename)  # Does type-checking of `filename`.
TypeError: expected str, bytes or os.PathLike object, not int
```

dshcherb replied:
Will move it to `_simulate_event` instead to make it closer to the use-case.

niemeyer commented on test/test_main.py:
```diff
@@ -130,34 +158,53 @@ def test_event_reemitted(self):
-            'mon_relation_changed': {'relation_name': 'mon', 'relation_id': 2, 'unit_name': 'remote/0'},
-            'ha_relation_broken': {'relation_name': 'ha', 'relation_id': 3},
-        }
+        events_under_test = (
```
This should be a list instead of a tuple.

Also, this layout tends to create better looking and more compact code:
```
some_list = [(
        foo,
        bar,
), (
        ... etc ...
)]
```

niemeyer commented on test/test_main.py:
```diff
@@ -43,6 +50,15 @@ class SymlinkTargetError(Exception):
     pass
 
 
+class SimulatedEventSpec:
```
I suggest s/Simulated//

`EventSpec` should be a fine name for this context, and will make the code below look slightly better.

niemeyer commented on test/test_main.py:
```diff
@@ -99,25 +115,37 @@ def _read_and_clear_state(self):
 
-    def _simulate_event(self, event_name, charm_config):
-        event_hook = JUJU_CHARM_DIR / f"hooks/{event_name.replace('_', '-')}"
+    def _simulate_event(self, event_spec, charm_config):
```
Looks like charm_config should be part of the event_spec.

dshcherb replied:
This introduces a requirement of passing it into every `EventSpec` constructor but it does seem to belong there.

niemeyer commented on test/test_model.py:
```diff
@@ -38,29 +39,35 @@ def relation_list(self, relation_id):
             raise op.model.RelationNotFound()
 
-    def relation_get(self, relation_id, member_name):
+    def relation_get(self, relation_id, member_name, app):
```
Same about app vs. boolean.

niemeyer commented on op/model.py:
```diff
@@ -287,9 +314,9 @@ def relation_list(self, relation_id):
+    def relation_get(self, relation_id, member_name, app):
         try:
-            return self._run('relation-get', '-r', str(relation_id), '-', member_name)
+            return self._run('relation-get', '-r', str(relation_id), '-', member_name, f'--app={app}')
```
Does it actually accept _"True"_ and _"False"_ as parameter values?

dshcherb replied:
Yes, I checked prior to implementing:

```
relation-set x=1 --app=True ; echo $?
0
```

dshcherb replied:
We might have a bug in Juju itself though. For app myapp, unit myunit/0:

`relation-get - myapp`

Returns `ERROR permission denied` on 2.7 RC4.

I will do another check on a newer RC.

niemeyer commented on op/model.py:
```diff
@@ -299,9 +326,9 @@ def relation_get(self, relation_id, member_name):
                 raise
 
-    def relation_set(self, relation_id, key, value):
+    def relation_set(self, relation_id, key, value, app):
```
Same about app vs. boolean.

niemeyer commented on op/model.py:
```diff
@@ -194,26 +207,34 @@ def __getitem__(self, key):
 # mutable or not is controlled by the flag.
-class RelationUnitData(LazyMapping, MutableMapping):
-    def __init__(self, relation, unit, is_mutable, backend):
+class RelationEntityData(LazyMapping, MutableMapping):
```
I suggest we name this as _RelationDataContent_ and amke the _entity_ attribute below private. We encourage this to be seen as content inside the "relation data", instead of the currently prevalent idea of independent "unit data" or "application data".

niemeyer commented on op/model.py:
```diff
@@ -194,26 +207,34 @@ def __getitem__(self, key):
+        if isinstance(self.entity, Application):
+            self._is_app = True
+        else:
+            self._is_app = False
```
```
self._is_app = isinstance(entity, Application)
```

dshcherb replied:
Indeed, I wonder how I didn't notice this.

dshcherb commented on op/model.py:
```diff
@@ -258,10 +279,16 @@ class RelationNotFound(ModelError):
 
 class ModelBackend:
+
+    LEASE_REFRESH_PERIOD = 30
```
Moved is-leader caching to a separate PR and fixed it there:

https://github.com/canonical/operator/pull/59

Going to drop this commit from here.

niemeyer commented on op/charm.py:
```diff
@@ -37,40 +37,52 @@ class LeaderSettingsChangedEvent(HookEvent):
-        super().__init__(handle, relation)
-        self.unit = unit
+        if self.unit:
+            snapshot['unit_name'] = self.unit.name
```
One more trivial nitpick: this function will likely look better with all of the empty lines dropped.

dshcherb commented on op/model.py:
```diff
@@ -194,26 +207,34 @@ def __getitem__(self, key):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
+            # This might happen due to a mid-hook leadership change besides trying to set data in remote data bags.
```
Agreed.

dshcherb commented:
Addressed the review comments in 13c16f5. Need to find a way to test the functionality of both using JUJU_REMOTE_APP and extracting a remote app name from JUJU_REMOTE_UNIT. This will require modifying `_simulate_event` in the test to take both scenarios into account in a controllable way.

niemeyer commented:
Is this ready for re-review?  The last comment above makes it look like it's still waiting for your action.

niemeyer commented on test/test_main.py:
```diff
@@ -132,10 +132,10 @@ def _simulate_event(self, event_spec):
+            # A remote app may not be in an event spec and so JUJU_REMOTE_APP will not be populated
+            # but it may still be a relation event with JUJU_REMOTE_UNIT for Juju < 2.7.
+            if remote_app is not None:
+                env['JUJU_REMOTE_APP'] = remote_app
```
The first sentence is the code stated with English, which is not useful. Let's please just drop the whole comment and have a line right above the env line with:
```
# For juju < 2.7 app name is extracted from JUJU_REMOTE_UNIT.
```

The second sentence is completely unrelated to the code here. It's the test that verifies such conditions, not the simulation logic. I suggest keeping the code and dropping the comment.

niemeyer commented on test/test_main.py:
```diff
@@ -206,6 +206,19 @@ def test_multiple_events_handled(self):
+        ), (
+            # Events without a remote app specified (for Juju < 2.7).
+            EventSpec(RelationJoinedEvent, 'db_relation_joined', relation_id=1,
+                      remote_app=None, remote_unit='remote/0', charm_config=charm_config),
```
The None is the default for all these cases.

niemeyer commented on op/main.py:
```diff
@@ -116,10 +116,25 @@ def _get_event_args(charm, bound_event):
+                remote_unit = model.get_unit(remote_unit_name)
+                return [relation, remote_app, remote_unit], {}
+            else:
+                return [relation, remote_app], {}
```
The logic above seems quite complicated, going back and forth between the two variables several times, but it looks like the conditions are much more strict than these branches make it sound.

Either we have a unit or we don't. That's about it. So this ought to be much simpler:
```
remote_app_name = os.environ.get('JUJU_REMOTE_APP', '')
remote_unit_name = os.environ.get('JUJU_REMOTE_UNIT', '')
if remote_app_name or remote_unit_name:
        if not remote_app_name:
                if '/' not in remote_unit_name:
                        raise RuntimeError(f'invalid remote unit name: {remote_unit_name}')
                remote_app_name = remote_unit_name.split('/')[0]
        args = [relation, model.get_app(remote_app_name)]
        if remote_unit_name:
                args.append(model.get_unit(remote_unit_name))
        return args, {}
return [], {}
```
The exception type is incorrect, but we can tune that later.

dshcherb replied:
OK, there is one missing bit in the above which is handling of `broken` relations - a relation needs to be returned without any units or applications present in that case. I will add it to my commit.

niemeyer commented on op/model.py:
```diff
@@ -186,8 +188,17 @@ def __repr__(self):
         self.relation = weakref.proxy(relation)
-        self._data = {our_unit: RelationUnitData(self.relation, our_unit, True, backend)}
-        self._data.update({unit: RelationUnitData(self.relation, unit, False, backend) for unit in self.relation.units})
+        self._data = {our_unit: RelationDataContent(self.relation, our_unit, (lambda _: True), backend)}
```
The lambda here escaped me earlier. This looks a bit strange at first sight.

dshcherb replied:
+1, answered in https://github.com/canonical/operator/pull/47#discussion_r352296886

niemeyer commented on op/model.py:
```diff
@@ -186,8 +188,17 @@ def __repr__(self):
+
+        # Whether the application data bag is mutable or not depends on whether this unit is a leader or not,
+        # but this is not guaranteed to be always true during the same hook execution.
+        self._data.update({our_unit.app: RelationDataContent(self.relation, our_unit.app, (lambda backend: backend.is_leader()), backend)})
```
Same.. this doesn't look clean at all. Note the hint of how the documentation is talking about the details which are not relevant in this place. RelationData shouldn't care about such things in its constructor ideally.

dshcherb replied:
+1, answered in https://github.com/canonical/operator/pull/47#discussion_r352296886

niemeyer commented on op/model.py:
```diff
@@ -204,26 +215,29 @@ def __getitem__(self, key):
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
```
Sorry for not noticing this earlier. This is outsourcing local knowledge from this type into the constructor. That's what's creating the awkwardness mentioned earlier. We need to move the logic into this type instead of putting everything into the constructor, including a lambda which is really just logic pertaining to this type.

dshcherb replied:
Agreed, there is a better way to do it now that we have `unit_name` and `app_name` attributes available from the `ModelBackend` type. Previously this was only accessible through the model which we do not have a reference to.

niemeyer commented on op/model.py:
```diff
@@ -336,17 +350,23 @@ def relation_list(self, relation_id):
+
         try:
-            return self._run('relation-get', '-r', str(relation_id), '-', member_name, return_output=True, use_json=True)
+            return self._run('relation-get', '-r', str(relation_id), '-', member_name, f'--app={is_app}', return_output=True, use_json=True)
```
Hmm... is it --app, or --application?  I recall seeing it both ways. In the spec I recall suggesting --app.

dshcherb replied:
It's `--app`.

https://github.com/juju/juju/blob/juju-2.7.0/worker/uniter/runner/jujuc/relation-get.go#L78-L81

dshcherb commented:
Thank you for the review, I added what I think should address the comments in https://github.com/canonical/operator/pull/47/commits/9ef78d10bbc081ac5b591a21110f4ebf1d8d4c41
